File: src/ai/agents/task-decomposer.ts
import { v4 as uuidv4 } from "uuid";
import { SubTask } from "../graph/state";

export interface DependencySubTask extends SubTask {
  order: number;
  dependsOnSubTaskId?: string;
}

export class TaskDecomposer {
  public static decomposeGoal(
    goalText: string,
    subTaskInputs: Array<{
      id?: string;
      type: "KNOWLEDGE_LOOKUP" | "DATABASE_QUERY" | "WORKFLOW_EXECUTION" | "DOCUMENT_DELIVERY";
      query: string;
      dependsOn?: string[];
      graphTraversal?: boolean;
      outputKey?: string;
    }>
  ): DependencySubTask[] {
    const subTasks: DependencySubTask[] = [];
    let previousId: string | undefined = undefined;

    subTaskInputs.slice(0, 10).forEach((input, index) => {
      const currentId = input.id || uuidv4();

      // If explicit dependsOn is provided, use it; otherwise fallback to linear previousId
      const dependsOn = input.dependsOn
        ? input.dependsOn
        : previousId
        ? [previousId]
        : [];

      subTasks.push({
        id: currentId,
        order: index + 1,
        type: input.type,
        query: input.query.trim(),
        dependsOn,
        dependsOnSubTaskId: previousId,
        status: "pending",
        graphTraversal: input.graphTraversal ?? false,
        ...(input.outputKey ? { outputKey: input.outputKey } : {}),
      });

      previousId = currentId;
    });

    console.log(`[TaskDecomposer] Compiled ${subTasks.length} sub-tasks for goal: "${goalText}"`);
    return subTasks;
  }
}

File: services/python-agent-service/agents/task_decomposer.py
import uuid
from typing import List, Dict, Any, Optional


class PythonTaskDecomposer:
    @classmethod
    def decompose_goal(
        cls,
        goal_text: str,
        raw_subtasks: List[Dict[str, Any]]
    ) -> List[Dict[str, Any]]:
        subtasks = []
        previous_id: Optional[str] = None

        for index, item in enumerate(raw_subtasks[:5]):
            current_id = str(uuid.uuid4())
            subtask = {
                "id": current_id,
                "order": index + 1,
                "type": item.get("type", "KNOWLEDGE_LOOKUP"),
                "query": item.get("query", goal_text).strip(),
                "dependsOnSubTaskId": previous_id,
                "status": "pending"
            }
            subtasks.append(subtask)
            previous_id = current_id

        print(f"[PythonTaskDecomposer] Decomposed goal '{goal_text}' into {len(subtasks)} ordered sub-tasks.")
        return subtasks
